clusterloader2/pkg/provider/provider.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"fmt"
+	"strings"
+)
+
+type InitOptions struct {
+	KubemarkRootKubeConfigPath string
+	ProviderName               string
+	ProviderConfigs            []string
+}
+
+type Features struct {
+	// SupportWindowsNodeScraping determines wheter scraping windows node in supported.
+	SupportWindowsNodeScraping bool
+	// SupportProbe determines whether probe is supported.
+	SupportProbe bool
+	// SupportImagePreload determines whether image preloading is supported.
+	SupportImagePreload bool
+	// SupportEnablePrometheusServer determines whether enabling prometheus server is possible.
+	SupportEnablePrometheusServer bool
+	// SupportSSHToMaster determines whether SSH access to master machines is possible.
+	// If false (impossible for many  providers), ClusterLoader will skip operations requiring it.
+	SupportSSHToMaster bool
+	// SupportAccessAPIServerPprofEndpoint determines whether accessing api server pprof endpoint is possible.
+	SupportAccessAPIServerPprofEndpoint bool
+	// SupportSnapshotPrometheusDisk determines whether snapshot prometheus disk is supported.
+	SupportSnapshotPrometheusDisk bool
+	// SupportNodeKiller determines whether node killer is supported.
+	SupportNodeKiller bool
+	// SupportGrabMetricsFromKubelets determines whether getting metrics from kubelet is supported.
+	SupportGrabMetricsFromKubelets bool
+
+	// ShouldPrometheusScrapeApiserverOnly determines if we should set PROMETHEUS_SCRAPE_APISERVER_ONLY by default.
+	ShouldPrometheusScrapeApiserverOnly bool
+}
+
+type Provider interface {
+	// Name returns name of this provider. It should used only in logs.
+	Name() string
+	// Features returns the feature supported by this provider.
+	Features() *Features
+
+	GetPortForComponent(componentName string) (int, error)
+	GetProtocolForComponent(componentName string) string
+	GetRootFrameworkKubeConfigOverride() string
+
+	RunSSHCommand(cmd, host string) (stdout string, stderr string, exit int, err error)
+}
+
+func NewProvider(initOptions *InitOptions) (Provider, error) {
+	configs := parseConfigs(initOptions.ProviderConfigs)
+	if initOptions.KubemarkRootKubeConfigPath != "" {
+		configs[RootKubeConfigKey] = initOptions.KubemarkRootKubeConfigPath
+	}
+	switch initOptions.ProviderName {
+	case "aks":
```

Comment:
Please make those constants, e.g.

AKSName = "aks"
AWSName  = "aws"
...

Reply:
And use them here and in individual providers.

Probably put them into dedciate constants.go file or sth like that.